File: app/src/main/java/im/point/dotty/network/PointAPI.java
package im.point.dotty.network;

import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface PointAPI {
    String BASE = "https://point.im";

    @GET("/api/recent")
    Call<PostsReply> getRecent(@Header("Authorization") String token, @Query("before") String before);

    @GET("/api/all")
    Call<PostsReply> getAll(@Header("Authorization") String token, @Query("before") String before);

    @GET("/api/comments")
    Call<PostsReply> getComments(@Header("Authorization") String token, @Query("before") String before);

    @GET("/api/bookmarks")
    Call<PostsReply> getBookmarks(@Header("Authorization") String token, @Query("before") String before);

    @GET("/api/tags")
    Call<PostsReply> getTagged(@Header("Authorization") String token, @Query("tag")String tag, @Query("before") String before);

    @GET("/api/tags/{user}")
    Call<PostsReply> getUserTagged(@Header("Authorization") String token, @Path("user") String user, @Query("tag")String tag, @Query("before") String before);

    @GET("/api/blog/{user}")
    Call<PostsReply> getUserPosts(@Header("Authorization") String token, @Path("user") String user, @Query("before") String before);

    @GET("/api/unread-counters")
    Call<CountersReply> getCounters(@Header("Authorization") String token);

    @GET("/api/post/{post}")
    Call<PostReply> getPost(@Header("Authorization") String token, @Path("post") String id);

    @GET("/api/user/login/{login}")
    Call<UserReply> getUser(@Header("Authorization") String token, @Path("login") String login);

    @GET("/api/user/id/{id}")
    Call<UserReply> getUser(@Header("Authorization") String token, @Path("id") int id);

    @GET("/api/me")
    Call<UserReply> getMyself(@Header("Authorization") String token);
}

File: app/src/main/java/im/point/dotty/network/PointAPI.kt
/*
 * Copyright (c) 2019-2021 Pavel A. Sokolov
 */
package im.point.dotty.network

import retrofit2.http.*

interface PointAPI {
    @GET("/api/recent")
    suspend fun getRecent(@Query("before") before: Long?): PostsReply

    @GET("/api/all")
    suspend fun getAll(@Query("before") before: Long?): PostsReply

    @GET("/api/comments")
    suspend fun getComments(@Query("before") before: Long?): PostsReply

    @GET("/api/bookmarks")
    suspend fun getBookmarks(@Query("before") before: Long?): PostsReply

    @GET("/api/tags")
    suspend fun getTagged(@Query("tag") tag: String, @Query("before") before: Long?): PostsReply

    @GET("/api/tags/{user}")
    suspend fun getUserTagged(
        @Path("user") user: String,
        @Query("tag") tag: String,
        @Query("before") before: Long?
    ): PostsReply

    @GET("/api/tags/{user}")
    suspend fun getUserTags(@Path("user") user: String): List<UserTagEntry>

    @GET("/api/blog/{user}")
    suspend fun getUserPosts(@Path("user") user: String, @Query("before") before: Long?): PostsReply

    @GET("/api/post/{post}")
    suspend fun getPost(@Path("post") id: String): PostReply

    @DELETE("/api/post/{post}")
    suspend fun deletePost(@Path("post") id: String): Envelope

    @GET("/api/user/login/{login}")
    suspend fun getUser(@Path("login") login: String): UserReply

    @GET("/api/user/id/{id}")
    suspend fun getUser(@Path("id") id: Long): UserReply

    @GET("/api/me")
    suspend fun getMe(): UserReply

    @GET("/api/unread-counters")
    suspend fun getUnreadCounters(): UnreadCounters

    @FormUrlEncoded
    @POST("api/post/{post}/r")
    suspend fun recommendPost(
            @Path("post") id: String,
            @Field("text") text: String? = null
    ): Envelope

    @DELETE("api/post/{post}/r")
    suspend fun unrecommendPost(@Path("post") id: String): Envelope

    @POST("api/post/{post}/s")
    suspend fun subscribeToPost(@Path("post") id: String): Envelope

    @DELETE("api/post/{post}/s")
    suspend fun unsubscribeFromPost(@Path("post") id: String): Envelope

    @POST("api/post/{post}/b")
    suspend fun bookmarkPost(@Path("post") id: String): Envelope

    @DELETE("api/post/{post}/b")
    suspend fun unbookmarkPost(@Path("post") id: String): Envelope

    @FormUrlEncoded
    @POST("api/post/{post}/pin")
    suspend fun pinPost(@Path("post") id: String, @Field("text") text: String? = null): Envelope

    @DELETE("api/post/{post}/unpin")
    suspend fun unpinPost(@Path("post") id: String): Envelope

    @POST("api/user/s/{login}")
    suspend fun subscribeToUser(@Path("login") login: String): Envelope

    @DELETE("api/user/s/{login}")
    suspend fun unsubscribeFromUser(@Path("login") login: String): Envelope

    @POST("api/user/sr/{login}")
    suspend fun subscribeToUserRecommendations(@Path("login") login: String): Envelope

    @DELETE("api/user/sr/{login}")
    suspend fun unsubscribeFromUserRecommendations(@Path("login") login: String): Envelope

    @POST("api/user/bl/{login}")
    suspend fun blockUser(@Path("login") login: String): Envelope

    @DELETE("api/user/bl/{login}")
    suspend fun unblockUser(@Path("login") login: String): Envelope

    @FormUrlEncoded
    @POST("/api/post/{id}")
    suspend fun addComment(
            @Path("id") id: String,
            @Field("text") text: String,
            @Field("comment_id") relyTo: Int? = null,
    ): Envelope

    @FormUrlEncoded
    @PATCH("/api/post/{id}/{number}")
    suspend fun editComment(
            @Path("id") id: String,
            @Path("number") number: Int,
            @Field("text") text: String,
    ): EditCommentReply

    @DELETE("/api/post/{id}/{number}")
    suspend fun deleteComment(
            @Path("id") id: String,
            @Path("number") number: Int
    ): Envelope

    @FormUrlEncoded
    @POST("/api/post/{id}/{number}/r")
    suspend fun recommendComment(
            @Path("id") id: String,
            @Path("number") number: Int,
            @Field("text") text: String,
    ): Envelope
}